luajit: adjust defaults LUA_(C)PATH

teto opened the pull request:
Running in an ubuntu, I get:
```
root@a5319686430d:/# luajit -e "print(package.path)" ./?.lua;/usr/share/luajit-2.1.0-beta3/?.lua;/usr/local/share/lua/5.1/?.lua;/usr/local/share/lua/5.1/?/init.lua;/usr/share/lua/5.1/?.lua;/usr/share/lua/5.1/?/init.lua 
root@a5319686430d:/# luajit -e "print(package.cpath)" ./?.so;/usr/local/lib/lua/5.1/?.so;/usr/lib/x86_64-linux-gnu/lua/5.1/?.so;/usr/local/lib/lua/5.1/loadall.so
```

so I've adjusted the paths accordingly

TODO target staging


## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

lolbinarycat commented on pkgs/development/interpreters/luajit/default.nix:
```diff
@@ -79,9 +81,9 @@ stdenv.mkDerivation rec {
       echo -e '
         #undef  LUA_PATH_DEFAULT
-        #define LUA_PATH_DEFAULT "./share/lua/${luaversion}/?.lua;./?.lua;./?/init.lua"
+        #define LUA_PATH_DEFAULT "./?.lua;./share/lua/${luaversion}/?.lua;./share/lua/${luaversion}/?/init.lua;"
```
this drops the `./?/init.lua`, is that intentional?

also, can you check if your issue is solved by #289135 ?

teto replied:
see my first command where default luajit's lua_path doesn't contain it.
It can be seen here as well https://github.com/LuaJIT/LuaJIT/blob/0d313b243194a0b8d2399d8b549ca5a0ff234db5/src/luaconf.h#L23

lolbinarycat replied:
consulting with luajit devs, this is intentional, but actually `./?/init.lua` should *only* be present in lua 5.3

teto replied:
interesting, I guess I will check in ubuntu containers the actual values and have those values set in the interpreter rather than luaLib now that we have structured envs (or if you beat me to it, you seem knowledgeable and enthusiastic about the lua ecosystem ;) ).

lolbinarycat replied:
i'm quite knowledgeable about lua itself, but i'm not actually that in-touch with the ecosystem (for example, i haven't actually used luarocks for any of my projects)

lolbinarycat replied:
> structured envs

i looked this up and got zero results, it seems a lot of new nix features are still underdocumented, could you elaborate?

teto commented:
I havent pushed my latest changes so not to override the CI but turns out we had almost every LUA_PATH wrong.
I found a better way to generate a valid default LUA_PATH, aka patch LUA_ROOT instead of patching all the rests. It's much simpler and robust and revelead our values were quite bad ^^'' . 
My last problem is that now that I include `./?.lua` in the pattern, the setup hook includes every derivation in LUA_PATH because of
```
local topDir="${absPattern%%\?*}" 
```
Once this is fixed we are good to go.

@lolbinarycat unrelated to this PR but mentioning this in case you haven't noticed but the propagatedBuildInputs are not added to LUA_PATH as it should. loadFromPropagatedInputs is never called pkgs/development/interpreters/lua-5/wrap.sh . This will be my next task since it's now hitting me when developing  https://github.com/nvim-neorocks/rocks.nvim/

lolbinarycat commented:
@teto 


> My last problem is that now that I include `./?.lua` in the pattern, the setup hook includes every derivation in LUA_PATH 

you shouldn't need to include `./?.lua` in the pattern, since it should be in LUA_PATH_DEFAULT, not in $LUA_PATH, and certainly not in `LuaPathSearchPaths`. (these *should* be seperate, LUA_PATH_DEFAULT controls the default value of `$LUA_PATH`, while `LuaPathSearchPaths` controls what gets added to `$LUA_PATH` by the envHook.)

teto commented:
yes, this was badly worded, what I meant is now that `./?.lua` is part of LUA_PATH_DEFAULT

lolbinarycat commented:
1. see my latest issue for why LUA_PATH_DEFAULT should not be used for the envHook
2. as far as i can tell, `wrap.sh` is not part of the normal envHook, but instead is a separate thing. (it builds a lua interpreter with the given packages embedded into it, its not clear why this would be useful instead of just including those packages as dependencies, maybe it should be deprecated?)

lolbinarycat commented:
why are we even patching LUA_PATH_DEFAULT via luaconf.h? to remove the `/usr/...` prefixes?

luajit explicitly says that patching `luaconf.h` shouldn't be needed, yet we do it anyways.

teto committed:
lua: smarter/more correct patching

In order to have the 'reset' LUA_PATH (aka `;;`) work, and for purity
reasons(removing /usr paths) we(I) decided to patch the lua interpreters default LUA_PATH.
Turns out the interpreters have different defaults and the patch was too
coarse.
There is smarter patching that can be done via LUA_ROOT.

Also luajit doesn't need patching at all since LUA_ROOT is set via the
build system.

teto committed:
lua.tests: update golden values for default LUA_PATH

when LUA_PATH contains `;;`, it is replaced by `;LUA_PATH_DEFAULT;` in
lua 5.1:
Updated https://github.com/lua/lua/blob/69ea087dff1daba25a2000dfb8f1883c17545b7a/loadlib.c#L599
More recent versions might be smarter ?
https://github.com/lua/lua/blob/65b07dd53d7938a60112fc4473f5cad3473e3534/loadlib.c#L301

teto committed:
lua: update setup-hook to limit LUA_PATH size

now that the lua interpreters include working directories with `./?.lua`
in LUA_PATH, the current test includes every derivation which quickly
becomes unreadable and unpractical.
The test is adapted to add a folder only if it can find lua files in the
subfolder.

teto committed:
doc/lua: mention the patching of interpreters

explain why they are patched in nixpkgs.

teto commented:
> why are we even patching LUA_PATH_DEFAULT via luaconf.h? to remove the /usr/... prefixes?

Yes and also to have `;;` work correctly. Wrapping the interpreter wouldn't work for `;;`. I've added a mention in the manual.

> luajit explicitly says that patching luaconf.h shouldn't be needed, yet we do it anyways.

I've added LUA_ROOT to makeFlags after finding in the luajit source code  an `#ifdef LUA_ROOT` (contrary to puc lua that just defines it). It works ok but I wondered if it is even necessary, maybe it's just the default behavior.

While fixing luajit and updating the golden test, I also noticed that the LUA_PATH contains `share/lua/luajit-2.1/?.lua` which doesn't work well with the setup hook. I've quickly introduced LUA_PATH_PATTERN and redirected lua.pkgs.luaLib.LuaPathList to interpreter-specific versions to account for it but I am open to suggestions. For instance we could ditch de the LUA_PATH_PATTERN thing for now and just check for lua files in a derivation to add it to the LUA_PATH.

teto commented:
arf and I wanted to update the release notes but seeing you duplicating the work I hurried up and forgot. Quite tired now but interested in your feedback/test. nixpkgs-review is great for that.

lolbinarycat commented:
> Yes and also to have ;; work correctly. 

what do you mean by "correctly"

i'm not advocating for wrapping lua, i'm advocating for setting PREFIX for luajit and LUA_ROOT for lua5

we only need to patch it to find its own system libraries, other modules will be located by the envHook

lolbinarycat commented on pkgs/development/interpreters/lua-5/interpreter.nix:
```diff
@@ -42,34 +42,50 @@ stdenv.mkDerivation (finalAttrs:
   LuaCPathSearchPaths = luaPackages.luaLib.luaCPathList;
+
+  # hook
+  LUA_PATH_PATTERN = [
```
this is recreating LuaPathSearchPaths

lolbinarycat commented on pkgs/development/interpreters/lua-5/interpreter.nix:
```diff
@@ -42,34 +42,50 @@ stdenv.mkDerivation (finalAttrs:
+    "share/lua/${luaversion}/?.lua"
+    "share/lua/${luaversion}/?/init.lua"
+  ]
+  ++ lib.optionals (lib.versionAtLeast self.luaversion "5.3") [
+    "./?.lua"
+    "./?/init.lua"
+    "lib/lua/${luaversion}/?.lua"
+    "lib/lua/${luaversion}/?/init.lua"
+  ];
```
1. now that we're not overriding LUA_PATH_DEFAULT, we don't need to handle relative module paths in nix, since we aren't overriding the part of luaconf.h that already handles this
2. this is repeating the mistake i mentioned in #293832
3. this is wrong, lua 5.1 also has `./?.lua` in its path, just not `./?/init.lua`

lolbinarycat commented on pkgs/development/interpreters/lua-5/interpreter.nix:
```diff
@@ -42,34 +42,50 @@ stdenv.mkDerivation (finalAttrs:
   setupHook = luaPackages.lua-setup-hook
-    finalAttrs.LuaPathSearchPaths
-    finalAttrs.LuaCPathSearchPaths;
+    finalAttrs.LUA_PATH_PATTERN
+    finalAttrs.LUA_CPATH_PATTERN;
```
this makes #293832 worse

lolbinarycat replied:
resolved.

lolbinarycat commented on pkgs/development/interpreters/lua-5/tests/assert.sh:
```diff
@@ -3,19 +3,19 @@
 #     fail "It should have been but it wasn't to be"
 function fail() {
-    echo "$1"
+    echo -e "$1"
```
`echo -e` evaluates backslash escape sequences, did you mean to do `>&2` to print to stderr?

teto replied:
nope, I wanted to echo the newlines. I am not sure what's best actually, do several fail calls or keep the newline ?  anyway this patch made actually harder to look at teh differences because $1  and $2 now dont start on same line. 
What I would like best is to show inline diffs. I know git diff can be used outside repos, not sure about enabling inline diffs, wether it's worth or not.

lolbinarycat commented on pkgs/development/interpreters/luajit/default.nix:
```diff
@@ -62,28 +62,33 @@ let
 
   luaversion = "5.1";
 
+  LUA_PATH_PATTERN = [
+    "./?.lua"
+    "./share/lua/luajit-${finalAttrs.version}/?.lua"
+    "./share/lua/luajit-${finalAttrs.version}/?/init.lua"
+  ];
+
+  LUA_CPATH_PATTERN = [
+    "./?.so"
+    "./lib/lua/${finalAttrs.luaversion}/?.so"
+    "./lib/lua/${finalAttrs.luaversion}/loadall.so"
+  ];
```
this feels like the opposite of DRY

also, the way these variables are named makes it sound like they are part of the lua specification instead of being a nix implementation detail.

teto replied:
what is DRY ?

lolbinarycat replied:
don't repeat yourself

lolbinarycat replied:
resolved.

lolbinarycat commented on pkgs/development/lua-modules/lib.nix:
```diff
@@ -22,13 +22,8 @@ rec {
-  luaCPathList = [
-    "lib/lua/${lua.luaversion}/?.so"
-  ];
+  luaPathList = lua.LUA_PATH_PATTERN;
+  luaCPathList = lua.LUA_CPATH_PATTERN;
```
do we even know what there are used for?  at the very least there should be a comment explaining their purpose.

teto replied:
agreed. This helps discriminate what package to bring into LUA_PATH (otherwise LUA_PATH grows too big) but it was buggy and that commit didn't improve the situation. I've removed the commit, it was buggy and added to confusion. Let's do a better thing in a future PR, for instance just add derivation to LUA_PATH if derivation contains lua files.

lolbinarycat replied:
> This helps discriminate what package to bring into LUA_PATH

no, `lua*.LuaPathSearchPaths` does that, or more precisely, the arguments to luaPackages.lua-setup-hook.

> I've removed the commit, it was buggy and added to confusion.

what commit? this was already part of the functionality of nixpkgs.

> for instance just add derivation to LUA_PATH if derivation contains lua files.

this seems like a horrible idea.  you'd eventually end up with serious namespace pollution issues where you're importing the guts of some lua program into your local namespace.

there's a reason the envHook prefix system exists.

lolbinarycat commented on pkgs/development/interpreters/luajit/default.nix:
```diff
@@ -62,28 +62,33 @@ let
+    "./lib/lua/${finalAttrs.luaversion}/loadall.so"
+  ];
+
+  # to compare with the default value, run `luajit -e "print(package.path)"`
```
this only works if you have luajit installed through something that isn't nixpkgs

teto replied:
I removed it

lolbinarycat commented on pkgs/development/interpreters/luajit/default.nix:
```diff
@@ -62,28 +62,33 @@ let
+  ];
+
+  # to compare with the default value, run `luajit -e "print(package.path)"`
+  # we patch the defaults otherwise `;;` will revert to an FHS setting
```
we are no longer patching this here.

lolbinarycat commented on pkgs/development/interpreters/luajit/default.nix:
```diff
@@ -98,6 +103,7 @@ stdenv.mkDerivation rec {
     "DEFAULT_CC=cc"
     "CROSS=${stdenv.cc.targetPrefix}"
     "HOST_CC=${buildStdenv.cc}/bin/cc"
+    "LUA_ROOT=$out"
```
1. this is unnecessary, PREFIX is already set
2. makeFlags get quoted at the shell level, so this would have to be `$(out)`
3. i'm not sure it would do anything even if it used the correct syntax

teto replied:
I tested and it is not necessary, I removed it.

lolbinarycat replied:
i know it's not necessary because i didn't have it in my competing pull request and that worked fine

lolbinarycat reviewed (commented):
the reason i started my own PR is because i wanted something more minimalist, this one makes a lot of changes, often without good reason (such as renaming variables)

lolbinarycat commented:
> nixpkgs-review is great for that.

i can't run it on my system, both due to the network of doing a full clone of nixpkgs, and due to the memory load of nix-build.

teto commented on doc/languages-frameworks/lua.section.md:
```diff
@@ -17,6 +17,9 @@ The main package set contains aliases to these package sets, e.g.
 `lua5_2.pkgs`.
 
+Note that nixpkgs patches the non-luajit interpreters to avoid referring to
+`/usr` and have `;;` (the "reset" LUA_PATH pattern) work correctly.
```
I've replaced it with
```
Note that nixpkgs patches the non-luajit interpreters to avoid referring to
`/usr` and have `;;` (a [placeholder](https://www.lua.org/manual/5.1/manual.html#pdf-package.path) replaced with the default LUA_PATH) work correctly.
```

lolbinarycat replied:
resolved

teto committed:
doc: update release notes for 24.05

with the lua patching

vcunat commented:
Maybe you'd want something like this?
https://github.com/NixOS/aarch64-build-box?tab=readme-ov-file#want-access

vcunat commented:
It's 80-core with 256M RAM.  Great for checking heavier stuff.

teto commented:
> i can't run it on my system, both due to the network of doing a full clone of nixpkgs, and due to the memory load of nix-build.

what I usually do is running nixpkgs-review on master or unstable which is much more tractable (with staging being often broken too) and at the end I rebase. If possible I test one specific thing that could be impacted (breaking neovim piss off people so I am careful about that one).

@lolbinarycat I think you should be able to get an account on https://nix-community.org/community-builder/ to run nixpkgs-review .

If you agree with the PR , I will rebase on staging and merge towards staging. Remember it's not supposed to be perfect, but this fixes a huge mistake in how we patched the interpreters. We can patch the rest in another PR.

vcunat commented:
> what I usually do is running nixpkgs-review on master or unstable which is much more tractable (with staging being often broken too) and at the end I rebase.

Yes, but if the issue is with download of nixpkgs repo and RAM for evaluation, this won't help.

lolbinarycat commented:
@vcunat that looks nice, but since i've only been working on nixpkgs for around a month or so, i don't have a ton of confidence it will be added.

lolbinarycat commented on pkgs/development/interpreters/lua-5/hooks/setup-hook.sh:
```diff
@@ -16,13 +16,21 @@ addToLuaSearchPathWithCustomDelimiter() {
   # delete longest match starting from the lua placeholder '?'
   local topDir="${absPattern%%\?*}"
 
-  # export only if the folder exists else LUA_PATH/LUA_CPATH grow too large
-  if [[ ! -d "$topDir" ]]; then return; fi
```
why are you doing this? do you *want* the path to be full of nonexistent files?

lolbinarycat commented on pkgs/development/interpreters/lua-5/hooks/setup-hook.sh:
```diff
@@ -16,13 +16,21 @@ addToLuaSearchPathWithCustomDelimiter() {
 
-  export "${varName}=${!varName:+${!varName};}${absPattern}"
+  # export only if the folder contains lua files
+  shopt -s globstar
+  for _file in "$topDir"/**.lua; do
+    export "${varName}=${!varName:+${!varName};}${absPattern}"
+    return;
+  done
```
using `**` here seems like a bad idea as long `$topDir` could easily be a derivation's root

teto replied:
why is it a bad idea ? the purpose of that code is to filter out derivations that dont add anything to lua else you end up with big LUA_PATH.

teto replied:
the advantage of the glob is that we can remove the `LuaPathSearchPaths` repetition you mentioned at the cost of including derivations that may not contain lua files readable from the interpreter LUA_PATH. It's an ok tradeoff I would say

lolbinarycat replied:
apon further review of the code, i realized this isn't actually the problem, but instead the problem was with unprefixed paths being added to LUA_PATH_PATTERN, but since that has been taken out, this issue is now resolved.

lolbinarycat commented on pkgs/development/interpreters/lua-5/hooks/setup-hook.sh:
```diff
@@ -16,13 +16,21 @@ addToLuaSearchPathWithCustomDelimiter() {
+  # if the path variable has not yet been set, initialize it to ";;"
+  # this is a magic value that will be replaced by the default,
+  # allowing relative modules to be used even when there are system modules.
+  if [[ -v "${varName}" ]]; then export "${varName}=;;"; fi
```
also, this code is completely wrong, has been reverted, and i'm making a PR with a fixed version right now.

lolbinarycat commented:
to summarize the status of the PR:
1. possible incorrect usage of `echo -e`
2. copy-pasting my code without understanding it

i think this PR is actually starting to shape up fairly well, but it still has a few big problems to fix.

lolbinarycat commented:
> Please be kind. Assume good faith rather than strawmen. This was a draft and that bit of code was reverted because its author didn't understand it either so it didn't matter to me.

i'm sorry, i just don't know how else i am supposed to take it when i see my code with my code comment being committed under someone else's name.

i assumed good faith at first and gently pointed it out, but after several comments back and forth, that comment was completely ignored.

lolbinarycat commented on pkgs/development/interpreters/lua-5/hooks/setup-hook.sh:
```diff
@@ -27,7 +22,14 @@ addToLuaSearchPathWithCustomDelimiter() {
+
+  for _file in ${absPattern/\?/\*\*}; do
   export "${varName}=${!varName:+${!varName};}${absPattern}"
+    return;
```
i'm not super familiar with how `shopt` works, but if we return here, won't `shopt` stay set?

also, the indentation on this line doesn't quite match.

teto replied:
indeed ty. so other hooks dont seem to care about disabling it but I remember reading something about not enabling it by default so I tried to disable it. I 've disabled it before the return and fixed the indentation.

With this PR, the setup-hook is still broken as it doesn't respect the interpreter LUA_PATH but what we could do to finally fix it is to remove the nix LuaPathSearchPaths and generate it from the relative paths in the default LUA_PATH. i.e., in the hook replace:
`  for pattern in @luapathsearchpaths@; do`
with 
```
IFS=';'
patterns=()
read -a patterns <<< $(lua -e print(package.path))
# filter out absolute paths from patterns ?
for pattern in patterns; do
``` 
we could even save the resulting patterns into the lua derivation as nix artefact to speed up thing.

lolbinarycat replied:
> With this PR, the setup-hook is still broken as it doesn't respect the interpreter LUA_PATH

i'm pretty sure adding `;;` to LUA_PATH fixes that?  are you doing your testing on staging?

lolbinarycat commented:
@ofborg build lua.passthru.tests lua5_1.passthru.tests luajit.passthru.tests

lolbinarycat reviewed (approved):
everything looks good to me

lolbinarycat commented on pkgs/development/interpreters/lua-5/tests/default.nix:
```diff
@@ -18,29 +15,43 @@ let
+      echo "${golden_LUA_PATH}"
+      assertStringEqual "$generated" "${golden_LUA_PATH}"
+      export LUA_PATH=";;"
+      assertStringEqual "$generated" "${golden_LUA_PATH}"
```
```suggestion
      assertStringEqual "$generated" ";${golden_LUA_PATH};"
```

`;;` actually gets replaced with `;LUA_PATH_DEFAULT;` to make sure paths don't get blended together.

teto replied:
well spotted ! This makes sense. Rebasing on your fix showed the test was broken.

I've updated the commit as you suggested with this additional commit message:
```
    when LUA_PATH contains `;;`, it is replaced by `;LUA_PATH_DEFAULT;` in
    lua 5.1:
    Updated https://github.com/lua/lua/blob/69ea087dff1daba25a2000dfb8f1883c17545b7a/loadlib.c#L599
    More recent versions might be smarter ?
    https://github.com/lua/lua/blob/65b07dd53d7938a60112fc4473f5cad3473e3534/loadlib.c#L301
```

lolbinarycat replied:
hmm, if it differs per-version that might make the tests significantly more complicated.  maybe we could just strip leading and trailing `;` in lua via gsub?  we'll see what ofborg says first.

lolbinarycat reviewed (changes requested):
sorry but it looks like there's one more issue that needs fixing

lolbinarycat commented:
@ofborg build luajit_2_0.passthru.tests luajit_2_1.passthru.tests lua5_1.passthru.tests lua5_2.passthru.tests lua5_3.passthru.tests lua5_4.passthru.tests

teto commented:
thanks for not giving up with this old geezer and the good reviews ! Planning to merge tomorrow after CI passes.

What's next:
1. we might want to test the default LUA_CPATH as well ? 
2. actually fix  pkgs/development/interpreters/lua-5/hooks/setup-hook.sh to not use luapathsearchpaths but the relative paths from the default LUA_PATH
3. remove luaPathList (it is used in several modules to work around the bugs we had)
4. fix the wrapLuaProgram to actually wrap propagated lua dependencies. Should solve so many bugs.

lolbinarycat commented:
> thanks for not giving up with this old geezer and the good reviews ! Planning to merge tomorrow after CI passes.

of course! i just want my favorite programming language to have a functioning nixpkgs ecosystem :)

> What's next:
> 
>     1. we might want to test the default LUA_CPATH as well ?

yep, as soon as this gets merged i'm planning on writing a bunch more unit tests, namely making sure LUA_PATH gets set correctly when there are propogatedBuildInputs.

>     2. actually fix  pkgs/development/interpreters/lua-5/hooks/setup-hook.sh to not use luapathsearchpaths but the relative paths from the default LUA_PATH

this would actually be a regression!  these need to be kept separate for the simple reason that they are often different.

in lua 5.2 and below, only `./?.lua` is checked in terms of relative paths, but for system modules, both `.../?.lua` and
`.../?/init.lua` are checked.

>     3. remove luaPathList (it is used in several modules to work around the bugs we had)

deduplication is good, although we should make sure we don't break the `override` interface too much.

>     4. fix the wrapLuaProgram to actually wrap propagated lua dependencies. Should solve so many bugs.

yep, and add unit tests for that.

teto commented:
> of course! i just want my favorite programming language to have a functioning nixpkgs ecosystem :)

Looking forward to someone taking over the lua maintenance (after some "running-in" ofc, I hope google-translate did a good job on "running-in") ;) As you could tell I am not an avid lua user, I just maintained the ecosystem because I needed it for neovim.

> this would actually be a regression! these need to be kept separate for the simple reason that they are often different.
> in lua 5.2 and below, only ./?.lua is checked in terms of relative paths, but for system modules, both .../?.lua and .../?/init.lua are checked.

so maybe not the default LUA_PATH but right now the lua setup hook looks at the `luaPathList` for patterns, which doesn't contain `./?.lua` so derivations with that pattern are not included in the generated LUA_PATH.

My guess (going to sleep sry ^^') is that in:
`nix-shell -p lua5_2 lua5_2.somePackageWithALuaFileAtTheRoot`
lua wont be able to require the module at /nix/store/HASH-somePackageWithALuaFileAtTheRoot/toto.lua

Feel free to ping me on matrix if you think this helps clear misunderstandings, my nickname is the same.

lolbinarycat commented:
> Looking forward to someone taking over the lua maintenance (after some "running-in" ofc, I hope google-translate did a good job on "running-in") ;) As you could tell I am not an avid lua user, I just maintained the ecosystem because I needed it for neovim.

i probably will set myself as a lua maintainer after this PR gets merged, although getting access to the community builder would let me actually perform real tests on the lua ecosystem.

> My guess (going to sleep sry ^^') is that in:
nix-shell -p lua5_2 lua5_2.somePackageWithALuaFileAtTheRoot
lua wont be able to require the module at /nix/store/HASH-somePackageWithALuaFileAtTheRoot/toto.lua

as far as i know this is intended behavior.  derivations (other than source derivations) generally should not put their files in the derivation root.  if there are any derivations that do this, those derivations should be changed.

teto commented:
so lua starts getting smarter with lua5_4 and doesn't add separators if there is nothing before or after `;;`. I've updated the tests to distinguish luajit_2_0 and luajit_2_1. We'll eventually have to remove the leading `;` from golden_LUA_PATH when testing with LUA_PATH != `;;` but that's ok for now.

vcunat commented:
I see that some builds broke because of this PR (succeed when I revert it locally):
  `libpeas2` https://hydra.nixos.org/build/254049435/nixlog/3/tail
  `knot-resolver` https://hydra.nixos.org/build/254409026/nixlog/2/tail
  
_I haven't investigated why, so far._

lolbinarycat commented:
it seems that buildLuarockPackage puts files in /share/lua/5.1, even withing luajitPackages.  luajit doesn't check there by default, nix used to patch it to, but we just removed that behavior.

we need to either reinstitute that, or change where files are installed (the latter feels more in line with how other lua*Packages work, but would at the very least need a changelog entry)

lolbinarycat commented:
to clarify my last point, it seems odd that luajitPackages and lua51Packages install to the same dir, but lua52Packages has a desperate dir.

teto commented:
`nix run .#luajitPackages.luarocks -- config` shows luarocks installs it in 5.1 (lua_module_path) even though it recognized a luajit interpreter. I dont know what is the correct behavior here, we should check with hisham or luajit folks. I just found https://stackoverflow.com/questions/6803582/luajit-not-seeing-rocks-installed-by-luarocks with no answer.

teto commented:
Looks like luarocks doesn't special case luajit https://github.com/luarocks/luarocks/blob/c16bdbc4cbb98c560fb7e8d4895d4d84afca2eb9/src/luarocks/core/cfg.lua#L194 so unless the system changes the default config lua_modules_path, it will install in 5.1 . I can't find any ticket of someone using luajit+luarocks with this issue that someone must have had ? anyway time to sleep

teto commented:
@vcunat so I had a look at both builds. In knot-resolver logs, we see in the logs that LUA_PATH contains paths towards libraries but http. As the derivation was a bit convoluted, I chose to look at libpeas which looked simpler. There doesn't seem to be an issue with the lua hooks.
```
$ nix develop ~/nixpkgs#libpeas2
$ genericBuild
...
meson.build:404:2: ERROR: Problem encountered: Lua51 requested but failed to locate suitable Lua51 and LGI support
$ echo $LUA_PATH  
`;;;/nix/store/yn9ncrx9fhp9zqpqisqcq601v2wlgp5w-lua5.1-lgi-0.9.2-1/share/lua/5.1/?.lua;/nix/store/yn9ncrx9fhp9zqpqisqcq601v2wlgp5w-lua5.1-lgi-0.9.2-1/share/lua/5.1/?/init.lua` 
$ cd ../  # go to build/ 's parent directory
$ mesonConfigurePhase # try again with same outcoe
```
so the path looked correct (folders exist) ?
I am not familiar with meson at all. Is it possible that meson's check is faulty ? or at least doesn't cover our usecase ? In the knot's resolver logs we can see `/usr/local` paths which should not appear so I wonder if there isn't any any magic somewhere in meson or build scripts of those 2 packages.

vcunat commented:
Let me provide a trivial test case that regresses:
```
$ env NIX_PATH=nixpkgs=. nix-shell --pure -Q -p luajitPackages.lua luajitPackages.http
nix-shell$ luajit
> require('http.request')
stdin:1: module 'http.request' not found:
[...]
```

vcunat commented:
meson itself isn't able to do any lua-specific stuff.  Of course, scripts written in meson can invoke lua/luajit, etc.

vcunat commented:
> In the knot's resolver logs we can see /usr/local paths which should not appear

Feel free to patch luajit not to search such paths by default.  I don't think it's closely related to this regression, though.

teto commented:
thanks for the easy repro

seemed like I used bash `globstar` wrong (I had tested but not correctly apparently, http is a good test because there is no lua file directly in share/lua/5.1/ but only in its http subfolder), aka a `**.lua` pattern doesn't work, it has to be `**/*.lua`

If you adapt the for loop in pkgs/development/interpreters/lua-5/hooks/setup-hook.sh with
```
-  for _file in ${absPattern/\?/\*\*}; do
+ for _file in ${absPattern/\?/\*\*\/\*}; do
```
it works. Gonna send a patch tomorrow, so dont revert if you can ;)

vcunat commented:
I wasn't hasty because... touching the file causes a nontrivial rebuild.  This bug is in master now, BTW.  (because I don't like delaying security fixes too much)

teto commented:
Ty. Opened a fix with its test at https://github.com/NixOS/nixpkgs/pull/300905 (test fails before the fix). Rebuilding my config with it
